main.cpp

```diff
@@ -40,26 +40,265 @@ write 3 UDTs below that EACH have:
+    bool m_inSignalPath {false} ;
+
+    EnvelopeGenerator();
+
+    void setADSR(int attack = 0, int decay = 0, int sustain = 0, int release = 0);
+    void connect(bool inSignalPath);
+    void reset();
+};
+
+EnvelopeGenerator::EnvelopeGenerator():m_attack(0), m_decay(0), m_sustain(0), m_release(0)
+{
+
+}
+
+void EnvelopeGenerator::setADSR(int attack, int decay, int sustain, int release)
+{
+    m_attack = attack;
+    m_decay = decay;
+    m_sustain = sustain;
+    m_release = release;
+
+    for (int c= 0; c <10; c++)
+        m_sustain = c;
+}
 
+void EnvelopeGenerator::connect(bool inSignalPath)
+{
+    m_inSignalPath = inSignalPath;
+}
+
+void EnvelopeGenerator::reset()
+{
+    m_attack = m_decay = m_sustain = m_release = 0;
+}
 /*
  UDT 2:
  */
+class LowFreqOscillator
+{
+public:
+    double  m_frequency;
+    int     m_waveform; 
+    int     m_modDepth;
+    bool    m_tapTempo {false};
+    bool    m_isActive {false};
+
+    LowFreqOscillator();
 
+    void enable();
+    void disable();
+    void setParams(double frequency, int waveform, int modDepth); 
+};
+
+LowFreqOscillator::LowFreqOscillator():m_frequency(440),m_waveform(0), m_modDepth(0)
+{
+    for (int c = 10; c < 100; ++c)
+    {   
+        m_frequency = c;
+    }
+
+    m_tapTempo = false;
+}
+
+void LowFreqOscillator::enable()
+{
+    m_isActive = true;
+}
+
+void LowFreqOscillator::disable()
+{
+    m_isActive = false;
+}
+
+void LowFreqOscillator::setParams(double frequency, int waveform, int modDepth)
+{
+    m_frequency = frequency;
+    m_waveform = waveform;
+    m_modDepth = modDepth;
+}
 /*
  UDT 3:
  */
+class Voice
+{
+public:
+    Voice();
+
+    void initialize();
+    void playSound();
+    void mute();
+
+    class Oscillator
+    {
+    public:
+        Oscillator();
+
+        void setFrequency(float freq);
+        void setWaveform(int freq);
+        void setOscType(int type);
+
+        EnvelopeGenerator m_ADSR;
+        float m_frequency;
+        int m_oscType, m_waveform;
+        bool m_sync;
+    };
+    
+    Oscillator m_Osc1, m_Osc2, m_Osc3, m_Osc4;
+    bool m_mute {true};
+};
+
+Voice::Voice():m_mute(true)
+{
+    initialize();
+}
+
+void Voice::initialize()
+{
+    m_Osc1.setWaveform(1);
+    m_Osc2.setWaveform(2);
+    m_Osc3.setWaveform(1);
+    m_Osc4.setWaveform(2);
+}
+
+void Voice::playSound()
+{
+    // play C Maj 7
+    m_Osc1.setFrequency(130.813f);
+    m_Osc2.setFrequency(164.814f);
+    m_Osc3.setFrequency(195.998f);
+    m_Osc3.setFrequency(246.942f);
+}
+
+void Voice::mute()
+{
+    m_mute = true;
+}
+
+Voice::Oscillator::Oscillator():m_oscType(1),m_sync(false)
+{
+    for (int j = 0; j< 10; j++)
+        m_sync = true;
+
+    m_ADSR.setADSR(0,0,0,0);
+}
 
+void Voice::Oscillator::setFrequency(float freq)
+{
+    m_frequency = freq;
+}
+
+void Voice::Oscillator::setWaveform(int waveform)
+{
+    m_waveform = waveform;
+}
 /*
  new UDT 4:
  */
+class synthMoog
+{
+public:
+    Voice m_Voice[4];
```

Comment:
Don't use raw arrays.  
Containers haven't been covered yet in the course. 
however, for the purpose of this exercise you can use `std::array<>` as that is the C++ equivalent of a raw C array.
the difference is that you have safety in the form of exceptions when you access indexes that are outside the bounds of the array.

put `#include <array>` at the top of the file and use `std::array<Voice, 4>`
- [ ] revise